server/src/main/java/org/elasticsearch/features/FeatureService.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.features;
+
+import org.elasticsearch.Version;
+import org.elasticsearch.common.Strings;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.NavigableMap;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.TreeSet;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Service responsible for registering features this node should publish to other nodes
+ */
+public class FeatureService {
+
+    private static final int CURRENT_ERA = Version.CURRENT.major;
+
+    /**
+     * The set of features that are implied by historical node versions.
+     */
+    private static final Map<Version, Set<String>> HISTORICAL_FEATURES = new ConcurrentHashMap<>();
+
+    /**
+     * Adds a feature that should be inferred when a node is at or above the specified historical version.
+     */
+    public static void registerHistoricalFeature(NodeFeature feature, Version version) {
+        if (version.after(Version.V_8_11_0)) {
+            throw new IllegalArgumentException("This method should only be used for historical features");
+        }
+        if (feature.era() != version.major) {
+            throw new IllegalArgumentException(Strings.format("Incorrect feature era %s for version %s", feature.era(), version.major));
+        }
+
+        // make this method idempotent (so it can be called multiple times in the same JVM for tests)
+        if (HISTORICAL_FEATURES.getOrDefault(version, Set.of()).contains(feature.id())) {
```

Comment:
This won't work in the case a `Node` is created & started, then a plugin is added and another `Node` is created - as the feature set will have been locked by then. You would have to register all the plugins before the first `Node` is created

Reply:
This is a testing scenario or something that could really happen? It means that 2 `Node` instances must be created inside the same JVM, right?

Reply:
Yup, it happens with some IT tests - but all plugins would be the same in that case